File: src/uintvar.c
#include "headers/uintvar.h"

bool is_uintvar_done(uint8_t byte){
    return (byte & 0x80) == 0;
}

unsigned int uintvar_to_int(const uint8_t *number) {
    unsigned int result = 0;
    int shift = 0;
    int data_bytes = 0; // To count how many bytes were successfully read

    for (int i = 0; i < 5; i++) { // Get the length of the uintvar
        data_bytes++;
        if (number[i] < 0x80) { // If the number is to large
            break;
        }
        if (i == 4) {
            perror("The uintvar is too large for this program. More than 32 bits");
            break; 
        }
    }

    shift = (data_bytes - 1) * 7; 
    for (int i = 0; i < data_bytes; i++) {
        uint8_t byte = number[i];
        result |= (unsigned int)(byte & 0x7F) << shift; //Append result
        shift -= 7; //Shift to the next byte
    }
    return result;
}

size_t int_to_uintvar(unsigned int value, uint8_t *number) {
    // handle zero explicitly
    if (value == 0) {
        number[0] = 0x00;
        return 1;
    }

    // collect 7-bit chunks from LSB to MSB into a small buffer
    uint8_t tmp[5];
    int count = 0;
    while (value != 0 && count < 5) {
        tmp[count++] = (uint8_t)(value & 0x7F);
        value >>= 7;
    }

    // if value still non-zero here, it exceeded 35 bits — but for 32-bit unsigned this won't happen
    // now write bytes in big-endian order: most-significant chunk first
    for (int i = count - 1; i >= 0; --i) {
        uint8_t b = tmp[i];
        if (i != 0) { // more bytes follow -> set continuation bit
            b |= 0x80;
        }
        *number++ = b;
    }

    return (size_t)count;
}

File: GO_Version/uintvar.go
package main

import "fmt"

func is_uintvar_done(value uint8) (bool) {
	return (value & 0b10000000) == 0;
}

func uintvar_to_uint(number []uint8) (uint32) {
	var result uint32 = 0;
	var data_bytes int = 0;

	for i := 0; i < 5; i++ {
		data_bytes ++;
		if is_uintvar_done(number[i]) {
			break;
		}
		if i == 4{
			fmt.Println("The uintvar is too large for this program. More than 32 bits")
			return 0;
		}
	}

	var shift int = (data_bytes - 1) * 7;
	for i := 0; i < data_bytes; i++ {
		var byte uint8 = number[i];
		result |= (uint32)(byte & 0x7F) << shift
		shift -= 7;
	}
	return result;
}

func uint_to_uintvar(value uint32) ([]uint8){
	if value == 0 {
		return []uint8{0};
	}

	var tmp []uint8;
	var count int = 0;
	for (value != 0) && count < 5 {
		count++;
		tmp = append(tmp, uint8(value & 0x7F));
		value >>= 7;
	}

	for i := count -1; i >= 0; i-- {
		var b uint8 = tmp[i];
		if i != 0 {
			b |= 0b10000000;
		}
		tmp[i] = b;
	}

	//Invert array.
	left, right := 0, len(tmp)-1
    for left < right {
        tmp[left], tmp[right] = tmp[right], tmp[left]
        left++
        right--
    }

	return tmp;
}